### DISCUSSION TITLE
Using multiple DBs?

### DISCUSSION THREAD

**Discussion by donn** [The-OpenROAD-Project/OpenROAD#4743] [General]
In issue #4692:

> If you use `db = odb.dbDatabase.create()` then you get a db with no Logger that can't issue messages. I suggest you just use the API that works.

The problem with that is while `odb.dbDatabase.create()` allows multiple DBs to exist simultaneously, the new API does not. We do have a number of scripts that operate on two databases simultaneously. Problem is, now we have three ways of doing this, none of which are ideal:

1. This fails with "No logger is installed in odb."

	```python3
	def _2r1w_db_old(db1_in, db2_in, db_out):
	    db1 = odb.dbDatabase.create()
	    db1 = odb.read_db(db1, db1_in)
	    db2 = odb.dbDatabase.create()
	    db2 = odb.read_db(db2, db2_in)
	    odb.write_db(db1, db_out)
	```

2. This fails with "You can't load a new db file as the db is already populated"

	```python3
	def _2r1w_db_new(db1_in, db2_in, db_out):
	    ord_tech1 = Tech()
	    design1 = Design(ord_tech1)
	    design1.readDb(db1_in)
	    
	    ord_tech2 = Tech()
	    design2 = Design(ord_tech2)
	    design2.readDb(db2_in)
	    
	    design1.writeDb(db_out)
	```

3. This works, but is horrifying and is likely to fail for multiple reasons in the future.

	```python3
	def _2r1w_db_mix(db1_in, db2_in, db_out):
	    ord_tech1 = Tech()
	    design1 = Design(ord_tech1)
	    design1.readDb(db1_in)
	    
	    db2 = odb.dbDatabase.create()
	    db2.setLogger(design1.getLogger())
	    db2 = odb.read_db(db2, db2_in)
	
	    design1.writeDb(db_out)
	```

Advice?

**Comment by maliberty** [comment]
I'm curious what your use case is for multiple dbs.  I guess you are only use odb APIs as nothing else in OR is likely to work in such a scenario.  

We could add an API to setup the logger.

**Reply by donn** [reply]
Correct, just ODB APIs- a couple things like comparing pins, grafting some instances from one design into another, etc.

The logger setup API would be appreciated.

**Comment by maliberty** [comment]
Does this satisfy your needs
```
from openroad import Design, Tech
import odb

ord_tech1 = Tech()
design1 = Design(ord_tech1)
design1.readDb('odb/test/data/design.odb')

db2 = Design.createDetachedDb()
db2 = odb.read_db(db2, 'odb/test/data/design.odb')

design1.writeDb('db1.odb')
odb.write_db(db2, 'db2.odb')
```

**Reply by maliberty** [reply]
Supporting multiple designs is a goal but it won't happen quickly.

**Reply by donn** [reply]
It would do, presuming all `odb` APIs still work

**Comment by maliberty** [comment]
Done #4749 

The odb APIs should be fine.
